linera-execution/src/runtime.rs

```diff
@@ -399,138 +509,168 @@ impl<const W: bool> SyncRuntimeInternal<W> {
     }
 }
 
-impl<const W: bool> SyncRuntime<W> {
-    fn new(runtime: SyncRuntimeInternal<W>) -> Self {
-        SyncRuntime(Arc::new(Mutex::new(runtime)))
+impl SyncRuntimeInternal<UserServiceInstance> {
+    fn load_service(
+        &mut self,
+        id: UserApplicationId,
+    ) -> Result<(UserServiceCode, UserApplicationDescription), ExecutionError> {
+        self.execution_state_sender
+            .send_request(|callback| Request::LoadService { id, callback })?
+            .recv_response()
+    }
+
+    /// Initializes a service instance with this runtime.
+    fn load_service_instance(
+        &mut self,
+        id: UserApplicationId,
+    ) -> Result<(Arc<Mutex<UserServiceInstance>>, Vec<u8>), ExecutionError> {
+        // Clippy's suggestion doesn't work because we need to borrow `self` while `entry` is still
+        // held
+        #[allow(clippy::map_entry)]
+        if !self.loaded_applications.contains_key(&id) {
+            let (code, description) = self.load_service(id)?;
+            let instance = code.instantiate(SyncRuntime(
+                self.reference
+                    .upgrade()
+                    .expect("`SyncRuntimeInner` should only be used by `SyncRuntime`"),
+            ))?;
+
+            self.loaded_applications
+                .insert(id, (Arc::new(Mutex::new(instance)), description));
+        }
+
+        let (instance, description) = self
+            .loaded_applications
+            .get(&id)
+            .expect("Application shoud be loaded");
+
+        Ok((instance.clone(), description.parameters.clone()))
     }
+}
 
-    fn into_inner(self) -> Option<SyncRuntimeInternal<W>> {
+impl<ContractOrService> SyncRuntime<ContractOrService> {
+    fn new(runtime: SyncRuntimeInternal<ContractOrService>) -> Self {
+        let mut this = SyncRuntime(Arc::new(Mutex::new(runtime)));
+        this.inner().reference = Arc::downgrade(&this.0);
+        this
+    }
+
+    fn into_inner(mut self) -> Option<SyncRuntimeInternal<ContractOrService>> {
+        self.inner().loaded_applications.clear();
```

Comment:
I'm not shocked by this at all but we may want to have this line moved at the callsite, next to the comment that you added anyway (or instead of the comment TBH).

Reply:
Done.